File: retrocpu_test_framework_ts/src/mn1613/main_stub.ts
import type { AsmCpuType } from "../types.js";

/**
 * MAIN が無いときの `_CODE` 原点（ワード）。MN1613 は割り込み領域を避ける。
 * @param cpu CPU
 * @param hasMain ソースに MAIN があるか
 * @returns スタブを置くワード。0 ならスタブ無し
 */
export function mn1613DefaultCodeOrgWord(
  cpu: AsmCpuType,
  hasMain: boolean,
): number {
  if (hasMain) {
    return 0;
  }
  if (cpu === "mn1613") {
    return 0x0200;
  }
  return 0;
}

/**
 * `_CODE` を指定ワードから始める MAIN スタブ。割り込み退避領域（0–7）と重ならないようにする。
 * @param orgWord `_CODE` 原点（ワード）
 * @param cpu 先頭 `.cpu` に書く CPU（既定 mn1613）
 * @returns スタブソース
 */
export function mn1613MainStub(
  orgWord: number,
  cpu: AsmCpuType = "mn1613",
): string {
  return [
    `\t.cpu\t${cpu}`,
    "\t.area\t_CODE\t\t(REL,CON)",
    `\t.org\t0x${orgWord.toString(16).toUpperCase()}`,
    "\t.global\t__TEST_FRAME_MAIN",
    "__TEST_FRAME_MAIN:",
    "\th",
    "",
  ].join("\n");
}

File: retrocpu_test_framework_rs/src/framework/mn1613/main_stub.rs
use crate::types::AsmCpuType;

pub fn mn1613_default_code_org_word(cpu: AsmCpuType, has_main: bool) -> u16 {
    if has_main {
        return 0;
    }
    match cpu {
        AsmCpuType::Mn1613 => 0x0200,
        AsmCpuType::Tms9995 => 0,
    }
}

pub fn mn1613_main_stub(org_word: u16, cpu: AsmCpuType) -> String {
    let cpu_name = match cpu {
        AsmCpuType::Mn1613 => "mn1613",
        AsmCpuType::Tms9995 => "tms9995",
    };

    format!(
        "\t.cpu\t{cpu_name}\n\t.area\t_CODE\t\t(REL,CON)\n\t.org\t0x{org_word:04X}\n\t.global\t__TEST_FRAME_MAIN\n__TEST_FRAME_MAIN:\n\th\n"
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_org_depends_on_main_and_cpu() {
        assert_eq!(
            mn1613_default_code_org_word(AsmCpuType::Mn1613, false),
            0x0200
        );
        assert_eq!(mn1613_default_code_org_word(AsmCpuType::Mn1613, true), 0);
        assert_eq!(mn1613_default_code_org_word(AsmCpuType::Tms9995, false), 0);
    }

    #[test]
    fn stub_contains_directives() {
        let s = mn1613_main_stub(0x0200, AsmCpuType::Mn1613);
        assert!(s.contains("\t.cpu\tmn1613"));
        assert!(s.contains("\t.org\t0x0200"));
        assert!(s.contains("__TEST_FRAME_MAIN:"));
    }
}
